src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -236,20 +199,55 @@ public Task WriteEndBatchAsync()
         }
 #endif
 
-        /// <summary>Starts a new changeset; can only be called after WriteStartBatch and if no other active operation or changeset exists.</summary>
+        /// <summary>
+        /// Starts a new changeset without specifying group id.
+        /// This can only be called after WriteStartBatch and if no other active operation or changeset exists.
+        /// </summary>
         public void WriteStartChangeset()
+        {
+            this.WriteStartChangeset(null);
+        }
+
+        /// <summary>
+        /// Starts a new changeset with the specified group id, which is null for multipart batch.
+        /// This can only be called after WriteStartBatch and if no other active operation or changeset exists.</summary>
+        /// <param name="groupId">
+        /// For Json batch, this is the atomic group id. Default value of null can be used for writing request and an GUID will be generated.
+        /// Non-null value from corresponding request atomic group should be used for writing response so that atomic groups
+        /// from request and response can be correlated.
+        /// For multipart batch, null value should be used and is ignored.
+        /// </param>
+        public void WriteStartChangeset(string groupId)
         {
             this.VerifyCanWriteStartChangeset(true);
-            this.WriteStartChangesetImplementation();
+            this.WriteStartChangesetImplementation(groupId);
+            this.UpdateAfterWriteStartChangeset();
         }
 
 #if PORTABLELIB
-        /// <summary>Asynchronously starts a new change set; can only be called after WriteStartBatch and if no other active operation or change set exists.</summary>
+        /// <summary>Asynchronously starts a new change set without specifying group id;
+        /// This can only be called after WriteStartBatch and if no other active operation or change set exists.</summary>
         /// <returns>A task instance that represents the asynchronous write operation.</returns>
         public Task WriteStartChangesetAsync()
+        {
+            return WriteStartChangesetAsync(null);

```

Comment:
As above, call WriteStartChangesetAsync(Guid.NewGuid().ToString()) so that downstream logic always has a non-null groupId.